File: get_ip/ifaddr.py
import sys
import socket
import fcntl
import struct

SIOCGIFADDR = 0x8915 # from bits/ioctls.h
IF_NAMESIZE = 16 # from net/if.h

sockaddr_in_fmt = "HHI"
# struct sockaddr_in is:
#    unsigned short (sa_family_t) "sin_family"
#    unsigned short "sin_port"
#    unsigned int   "sin_addr"

ifreq_fmt = "16s" + sockaddr_in_fmt
# struct ifreq is:
#    char ifname[16]
#    struct sockaddr_in

def get_ip(ifname):
    sk = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
    blob = struct.pack(ifreq_fmt, ifname, 0, 0, 0)
    r = fcntl.ioctl(sk.fileno(), SIOCGIFADDR, blob)
    ifreq = struct.unpack(ifreq_fmt, r)
    ip = socket.inet_ntoa(struct.pack("I", ifreq[3]))
    return ip

if __name__ == "__main__":
    print get_ip(sys.argv[1])

File: get_ip/ifaddr.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>

int main(int argc, char *argv[])
{
	struct sockaddr_in *sin;
	struct ifreq ifr;
	int sk;

	if (argc != 2) {
		fprintf(stderr, "USAGE: %s <interface>\n", argv[0]);
		exit(1);
	}

	sk = socket(AF_INET, SOCK_DGRAM, 0);
	if (sk < 0) {
		perror("socket()");
		exit(1);
	}

	memset(&ifr, 0, sizeof(ifr));
	ifr.ifr_addr.sa_family = AF_INET;
	strncpy(ifr.ifr_name, argv[1], sizeof(ifr.ifr_name));

	if (ioctl(sk, SIOCGIFADDR, &ifr) < 0) {
		perror("ioctl()");
		exit(1);
	}

	sin = (struct sockaddr_in *)&ifr.ifr_addr;
	printf("IP: %s\n", inet_ntoa(sin->sin_addr));

	return 0;
}
